File: src/main/scala/com/imcode/imcms/mapping/CacheWrapper.scala
package com.imcode
package imcms
package mapping

import _root_.net.sf.ehcache.{Cache, Element, Ehcache}
import _root_.net.sf.ehcache.config.CacheConfiguration

case class CacheWrapper[K >: Null, V >: Null](cache: Ehcache) {

  // Compiles, but Intellij can't infer
  // Option(cache.get(key)).map(_.getObjectValue).orNull.asInstanceOf[V]
  def get(key: K): V =
    cache.get(key) |> {
      case null => null
      case element => element.getObjectValue
    } |> { _.asInstanceOf[V] }

  def put(key: K, value: V): Unit = cache.put(new Element(key, value))

  def remove(key: K): Boolean = cache.remove(key)

  def getOrPut(key: K)(compute: => V): V =
    get(key) match {
      case value if value != null => value
      case _ => compute |>> { value => put(key, value) }
    }
}


object CacheWrapper {
  def apply[K >: Null, V >: Null](cacheConfiguration: CacheConfiguration) =
    new CacheWrapper[K, V](new Cache(cacheConfiguration))
}

File: src/main/java/com/imcode/imcms/mapping/CacheWrapper.java
package com.imcode.imcms.mapping;

import net.sf.ehcache.Cache;
import net.sf.ehcache.Ehcache;
import net.sf.ehcache.Element;
import net.sf.ehcache.config.CacheConfiguration;

import java.util.function.Supplier;

public class CacheWrapper<K, V> {

    private final Ehcache cache;

    public CacheWrapper(Ehcache cache) {
        this.cache = cache;
    }

    public static <K, V> CacheWrapper<K, V> of(CacheConfiguration cacheConfiguration) {
        return new CacheWrapper<>(new Cache(cacheConfiguration));
    }

    public V get(K key) {
        Element element = cache.get(key);
        return element == null ? null : (V) element.getObjectValue();
    }

    public void put(K key, V value) {
        cache.put(new Element(key, value));
    }

    public boolean remove(K key) {
        return cache.remove(key);
    }

    public V getOrPut(K key, Supplier<V> valueSupplier) {
        V value = get(key);
        if (value == null) {
            value = valueSupplier.get();
            put(key, value);
        }

        return value;
    }

    public Ehcache cache() {
        return cache;
    }
}
